File: src/python/p300top399/345.反转字符串中的元音字母.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   345.反转字符串中的元音字母.py
@Time    :   2020/08/30 10:01:34
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""

#
# @lc app=leetcode.cn id=345 lang=python3
#
# [345] 反转字符串中的元音字母
#
# https://leetcode-cn.com/problems/reverse-vowels-of-a-string/description/
#
# algorithms
# Easy (50.61%)
# Likes:    112
# Dislikes: 0
# Total Accepted:    45.4K
# Total Submissions: 89.8K
# Testcase Example:  '"hello"'
#
# 编写一个函数，以字符串作为输入，反转该字符串中的元音字母。
#
#
#
# 示例 1：
#
# 输入："hello"
# 输出："holle"
#
#
# 示例 2：
#
# 输入："leetcode"
# 输出："leotcede"
#
#
#
# 提示：
#
#
# 元音字母不包含字母 "y" 。
#
#
#


# @lc code=start
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowel = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        s = list(s)
        i, j = 0, len(s) - 1
        while i < j:
            while i < j and s[i] not in vowel:
                i += 1
            while i < j and s[j] not in vowel:
                j -= 1
            if i < j:
                s[i], s[j] = s[j], s[i]
                i += 1
                j -= 1
        return ''.join(s)


# @lc code=end

if __name__ == '__main__':
    solu = Solution()
    print(solu.reverseVowels('aA'))
    print(solu.reverseVowels('leetcode'))

File: src/golang/p300to399/345.反转字符串中的元音字母.go
package p300to399

/*
 * @lc app=leetcode.cn id=345 lang=golang
 *
 * [345] 反转字符串中的元音字母
 *
 * https://leetcode-cn.com/problems/reverse-vowels-of-a-string/description/
 *
 * algorithms
 * Easy (52.77%)
 * Likes:    197
 * Dislikes: 0
 * Total Accepted:    90.5K
 * Total Submissions: 171.6K
 * Testcase Example:  '"hello"'
 *
 * 给你一个字符串 s ，仅反转字符串中的所有元音字母，并返回结果字符串。
 *
 * 元音字母包括 'a'、'e'、'i'、'o'、'u'，且可能以大小写两种形式出现。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：s = "hello"
 * 输出："holle"
 *
 *
 * 示例 2：
 *
 *
 * 输入：s = "leetcode"
 * 输出："leotcede"
 *
 *
 *
 * 提示：
 *
 *
 * 1 <= s.length <= 3 * 10^5
 * s 由 可打印的 ASCII 字符组成
 *
 *
 */

// @lc code=start
func reverseVowels(s string) string {
	vowels := map[byte]struct{}{}
	for _, ch := range "aeiouAEIOU" {
		vowels[byte(ch)] = struct{}{}
	}

	bs := []byte(s)
	i, j := 0, len(bs)-1
	for i < j {
		for _, ok := vowels[bs[i]]; !ok && i < j; _, ok = vowels[bs[i]] {
			i++
		}
		for _, ok := vowels[bs[j]]; !ok && i < j; _, ok = vowels[bs[j]] {
			j--
		}
		if i < j {
			bs[i], bs[j] = bs[j], bs[i]
			i++
			j--
		}
	}

	return string(bs)
}

// @lc code=end
